qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,224 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, Optional, Set, Union
+
+import sympy
+from attrs import frozen
+
+from qualtran import Bloq, Register, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+from qualtran._infra.composite_bloq import BloqBuilder, SoquetT
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class ModAdd(Bloq):
```

Comment:
How is this different to the existing modular arithmetic bloqs?

Reply:
My main motivation for creating a separate new bloq is because a modular addition subroutine will need an input register for the modulus p which none of the existing bloqs currently have. Can be ignored if desired (and kinda is ignored in the paper), but I wanted to try and be more thorough.

Reply:
I think in general we'll need a way to distinguish quantum-quantum vs quantum-classical operations which have different costs.

However, I don't think this applies here. In fig 6 it's shown that x and y are the quantum registers (containing quantum data). The modulus is a classical parameter